### PR TITLE
Automated model registry

### PR DESCRIPTION
**Pull request by CharlelieLrt**
# PhysicsNeMo Pull Request

## Description

This PR introduces multiple changes aimed at standardizing some aspect of the model registry.

### Automatic registration

New subclasses of `physicsnemo.core.Module` will always be added to the registry by default, unless they are created with `_register=False`:

```python
from physicsnemo.core import Module

class MyModel(Module, _register=False)
    def __init__(self):
        ...
```
**Note:** the `_register` argument is for internal use only and is not meant to be exposed. 
⚠️ **Important :** this is a backward-compatibility breaking change that will most likely create multiple conflict, since names in the registry are unique.

### Naming consistency and flexibility in for models imported from PyTorch

A new optional `name` parameter in the method `Module.from_torch()`. This is aimed at giving flexibility for naming `Module` subclasses that are created from subclasses of `torch.nn.Module`.  For consistency, this `name` parameter is also used to register the model class to the registry. If not provided, the model class name will default to that of the PyTorch model class.

⚠️ **Important :** this is a backward-compatibility breaking change for the registry, since prior to that `from_torch` would add model classes to the registry with an *implicit* built-in name of the form `<name of the PyTorch class>PhysicsNeMoModel`. This built-in name was implicit and difficult to discover for users (it would typically require querying the registry to find out this name). The proposed changes remove this implicit behavior.

### Others

Improved documentation, type hints, and error messages in `module.py` and `registry.py`

## Checklist

- [ ] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/physicsnemo/blob/main/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.
- [ ] The [CHANGELOG.md](https://github.com/NVIDIA/physicsnemo/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/physicsnemo/issues) is linked to this pull request.

## Review Process

**All PRs are reviewed by the PhysicsNeMo team before merging.**

Depending on which files are changed, GitHub may automatically assign a maintainer for review.

We are also testing AI-based code review tools (e.g., Greptile), which may add automated comments with a confidence score.
This score reflects the AI’s assessment of merge readiness and is **not** a qualitative judgment of your work, nor is
it an indication that the PR will be accepted / rejected.

AI-generated feedback should be reviewed critically for usefulness.
You are not required to respond to every AI comment, but they are intended to help both authors and reviewers.
Please react to Greptile comments with 👍 or 👎 to provide feedback on their accuracy.

### CODE REVIEW

File: physicsnemo/core/module.py [hunk 1]
```diff
@@ -197,6 +196,22 @@ def __init__(self, meta: Union[ModelMetaData, None] = None):
...
+        Parameters
+        ----------
+        _register : bool, optional
+            For internal use only. Whether to register the subclass in the
+            model registry, by default True
+        """
+        super().__init_subclass__()
+        if _register:
+            registry = ModelRegistry()
+            registry.register(cls, cls.__name__)
```
[the remark this answers is not in the document]
Comment (CharlelieLrt) [reply physicsnemo/core/module.py:213]: All these comments are valid, and the backward-compatibility breaking changes are expected and documented in the description of this PR. The aforementioned tests should be adapted with different model names and/or clearing the registry between tests.

Defer registration until explicitly requested

That is also a valid point and another possible design strategy: instead of implicitly registering models, we could expose and document the `register` parameter. The same `register` parameter could also be added to the `Module.from_torch()` method. However, the registry is seen as a key feature for PhysicsdNeMo's Module, appreciated by many users. This motivates our decision to automate the registration by make this implicit registration the default behavior.

[the remark this answers is not in the document]
Comment (CharlelieLrt) [reply physicsnemo/core/module.py:213]: Fixed the unused `name` argument
